autocue.cue_file.liq

```diff
@@ -259,119 +246,97 @@ end
 
 # Deconstruct a SemVer version, return a record
 def semver(s) =
-  s = null.get(default="", s)
   # SemVer RegEx, see https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
   #r = r/(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/gm
-  r = r/(?P<major>0|[1-9]\d*)\.(?P<minor>0|[1-9]\d*)\.(?P<patch>0|[1-9]\d*)(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?P<build>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/gm
-  v = r.exec(s)
-  #print(v)
-  {
-    version = v[0],
-    major = v.groups["major"],
-    minor = v.groups["minor"],
-    patch = v.groups["patch"],
-    prerelease = v.groups["prerelease"],
-    build = v.groups["build"]
-  }
-end
+  r = r/(?<major>0|[1-9]\d*)\.(?<minor>0|[1-9]\d*)\.(?<patch>0|[1-9]\d*)(?:-(?<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?<build>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/gm
 
-# Compare two SemVers
-# The return value is negative if ver1 < ver2,
-# zero if ver1 == ver2 and strictly positive if ver1 > ver2
-def semver_compare(s1, s2) =
-  s1 = null.get(default="", s1)
-  s2 = null.get(default="", s2)
-  v1 = semver(s1)
-  v2 = semver(s2)
-
-  if v1.major == v2.major and v1.minor == v2.minor and v1.patch == v2.patch then
-    0
-  elsif v1.major > v2.major then
-    1
-  elsif v1.major >= v2.major and v1.minor > v2.minor then
-    1
-  elsif v1.major >= v2.major and v1.minor >= v2.minor and v1.patch > v2.patch then
-    1
+  if r.test(s) then
+    v = r.exec(s)
+    #print(v)
+    (
+      {
+        version=v[0],
+        major=int_of_string(v.groups["major"]),
+        minor=int_of_string(v.groups["minor"]),
+        patch=int_of_string(v.groups["patch"]),
+        prerelease=v.groups["prerelease"],
+        build=v.groups["build"]
+      }
+    :
+      {
+        version: string,
+        major?: int,
+        minor?: int,
+        patch?: int,
+        prerelease?: string,
+        build?: string
```

Comment:
These make it possible to return a default value when the regexp doesn't match where all optional values are not defined, see below.